### DISCUSSION TITLE
Full working example for AsyncAioPikaManager

### DISCUSSION THREAD

**Discussion by shay-te** [miguelgrinberg/python-socketio#785] [General]
thank you for reading this question.

when using `AsyncAioPikaManager` what server will be running? what is the complete example to run using asyncio.

can you please provide a complete example?

this is really confusing for me

thank you

**Comment by miguelgrinberg** [comment]
@shay-te You will be using the `AsyncServer()` class, and passing the instance of `AsyncAioPikaManager` in the `client_manager` argument. Something like this:

    sio = AsyncServer(client_manager=AsyncAioPikaManager())

You can take any of the async examples in this repository and add the Pika client manager to it. Nothing else changes, other than the fact that once you have Pika added to the mix you are able to run multiple servers behind a load balancer and they all communicate with each other.

**Reply by shay-te** [reply]
Thank you for the fast response. 

I think it's a bit confusing to me. 

just verifying 
Will use the simple `aiohttp` to communicate with the client. 

what if I wish to get a message from another service. can use the  `AsyncAioPikaManager` for that? 

thanks again

**Reply by miguelgrinberg** [reply]
The client managers serve two purposes: 1) allowing you to have multiple servers behind a load balancer and 2) allowing an external process to emit events to your Socket.IO clients. The communication between the clients and the server are not affected by the use of a client manager. The client manager is only used internally as an aid in implementing the two cases I listed above.

**Reply by shay-te** [reply]
Thank you Miguel
so the `client_manager` help me in the following scenario 
A = service with the main product
B = service with socket io
MEDIATOR = rabbitmq
 
if a case when I wish to notify the client from another service `A -> MEDIATOR -> B`
1. will this work? 
2. in this case I MEDIATOR events also will raise from the `@sio.event`?
3. how can I differentiate them from client messages? 

Thank you

**Reply by miguelgrinberg** [reply]
First of all, if you have a main web server and you want to add Socket.IO, then you can do this all within the same process. The Socket.IO server can attach to your existing application and both can run together, and that eliminates the need to use the message queue in the middle.

Now, if you do decide to work with the queue, the two things that you can do are the two that I mentioned above. In your situation A will be able to send notifications to connected clients. That's all you get. A cannot get events from clients, this is a send-only solution. If you want to receive notifications from clients in the server, the handlers need to be installed in B.

**Reply by shay-te** [reply]
Thank you your answer is helping a lot.
